### ISSUE TITLE
Documentation request, details on improving displacements.displacement results

### ISSUE THREAD

**Issue by gdevenyi** [m-pilia/disptools#1]
The documentation at:
https://martinopilia.com/disptools/#disptools.displacements.displacement

Describes all the parameters of the function, however, it isn't very clear on how to trade off computational time for accuracy.

I'm willing to spend lots of computation on a getting a very accurate result because I'm working on simulation-type work.

How would I go about tuning the parameters to achieve this?

**Comment by gdevenyi** [comment]
In particular, there's no clear relationship between the output during run and which parameters which adjust the thresholds for those values:
```
Iteration  2967:  total error 9.504962e-05  max voxel error 1.013279e-05  eta 2.484279e-01
```

**Comment by m-pilia** [comment]
Hi Gabriel! It is difficult to give a universal answer, since the algorithms can behave differently depending on the properties of the input. In general, the problem itself is not well posed, and I observed that it can be very difficult to converge to a good solution if the Jacobian has sharp variations or discontinuities.

First thing I would note is that there are two different algorithms available (`gradient` and `greedy`), and the algorithm can be selected by specifying the `algorithm` parameter when calling `displacements.displacement()`. The default algorithm is `gradient`, that has a more complicated cost function and in theory could be a bit more robust. On the other hand, the `greedy` algorithm has a simpler cost function and it generally behaves well, in practice I ended up preferring it in most cases. (For the sake of completeness, there is a third algorithm, `matching`, that I did not implement but comes from another project, and I used it mostly as a baseline for evaluation; it is *very* slow and has a very strong regularisation, so it may not converge after a certain point).

About the parameters, both methods use a Armijo condition to control the step length `eta` over the iterations, so if the goal is high accuracy I would suggest to not allow large values of `eta`, especially if the input Jacobian is not very smooth. Using a smaller value for `eta_max` and possibly decreasing `alpha` can be helpful in this scenario.

I would also note that the default build is single precision float, so that can be a limiting factor depending on what order of magnitude is desired for the precision of the result. I never needed to work in double precision with this package, but it is necessary to do so in order to go anywere below `1e-6`. It is possible to build the software in double precision by adding the build argument `-DDISPTOOLS_DOUBLE=ON`.

**Comment by gdevenyi** [comment]
ooh, single vs double is very important thanks :) will adjust that.

I would say my jacobian is probably very not-smooth, it's a segmented piece of anatomy I'm artificially increasing in size, I'll see that I can adjust with eta.

**Comment by m-pilia** [comment]
I see, I think it should work for this type of application. It may be more or less difficult depending on how far from 1.0 is the Jacobian in the segmented area, but if the values are not crazy then it should be doable.

One important thing I would recommend is to use a mask for the background whenever possible (the `mask` argument of `displacement()`). It is especially important if the input Jacobian does not preserve the total volume of the image, but it helps a lot in any case. That allows to redistribute the residual volume changes in the background of the image (hopefully it should not matter if the background of the image gets expanded or compressed), and it greatly helps to converge.

**Comment by gdevenyi** [comment]
Indeed, I am using a mask, and my largest jacobian value I'm testing is 20% change (0.8 and 1.2 respectively).

Here's a view for one subject:
![image](https://user-images.githubusercontent.com/3001850/60130818-0225c980-9766-11e9-9cac-3287f87afbbb.png)

**Comment by m-pilia** [comment]
Very interesting. Yes, I think 20% volume change on a region of that size should be doable.
